File: LC/[1278] Palindrome Partitioning III.py
class Solution:
    def palindromePartition(self, s: str, k: int) -> int:
        """
        - S: divide S into K 
        - k:
        - N: ?
        => min num of chars need to change
        * 无后效性
        dp[ii][kk]: the minimal number of characters that you need to change to divide the string s[:ii+1] into k substrings that are all palindromes

        * consider s[0:ii+1], divide this substring into k substring

        {xxxxxx}[jxxx ii] # => min(dp[jj-1][k-1] + helper(s[j:ii+1])) for jj=1,2,3...ii
        k-1     | last one                     ^^^^^^^^^^^^^^^
                                               required flips to make last interval palindrome

        for ii in range(1, N):
            for k in range(1, min(KK, ii+1))
                for jj in range(k, ii+1): # k intervals, jj is the start of the last interval
                    dp[ii][kk]=dp[jj-1][k-1] + helper(s, jj, ii)

        return dp[-1][-1]
        """
        s = '#'+s
        N = len(s)
        dp = [[math.inf]*(k+1) for _ in range(N)]
        dp[0][0] = 0

        memo = {}

        def helper(s, st, ed):
            if (st, ed) in memo:
                return memo[(st, ed)]
            ll, rr = st, ed
            cnt = 0
            while ll < rr:
                if s[ll] != s[rr]:
                    cnt += 1
                ll += 1
                rr -= 1
            memo[(st, ed)] = cnt
            return cnt

        for ii in range(1, N):
            for kk in range(1, min(k+1, ii+1)):
                for jj in range(kk, ii+1):  # k intervals, jj is the start of the last interval
                    # try cache helper for optimization
                    dp[ii][kk] = min(dp[ii][kk], dp[jj-1]
                                     [kk-1] + helper(s, jj, ii))

        return dp[-1][-1]

File: LC/[1278] Palindrome Partitioning III.java
class Solution {
    private int helper(int st, int ed, String s){
        int cnt = 0;
        while(st < ed){
            if(s.charAt(st) != s.charAt(ed)) ++cnt;
            ++st;
            --ed;
        }

        return cnt;
    }
    public int palindromePartition(String s, int k) {
        int N = s.length();
        int[][] dp = new int[N+1][k+1];
        for(int ii=0; ii<=N; ++ii){
            for(int jj=0; jj<=k; ++jj){
                dp[ii][jj] = Integer.MAX_VALUE/2;
            }
        }

        s = '#'+s; //java string concatenation
        dp[0][0] = 0;

        for(int ii=1; ii<=N; ++ii){
            for(int kk=1; kk <= Math.min(k, ii); ++kk){
                for(int jj=kk; jj<=ii; ++jj){
                    dp[ii][kk] = Math.min(dp[ii][kk], dp[jj-1][kk-1] + helper(jj, ii, s));
                }

            }

        }

        return dp[N][k];
    }
}

/*
Problem statements: given nums and k
    - partition nums in to k non-empty subarrays s.t. each subarray is an palindrome
    - you can change any char whenever necessary
    - the min num of changes you can make all k subarray palindrome

Analysis: 
    - partition array into k sub-array -> DP4
    - dp[ii][kk]: min ops required to make nums[:ii+1] partition into kk subarrays and each subarray is palindrome
    - dp[ii][kk] = min(dp[ii-1][kk-1], x+dp[jj-1][kk-1])
    - where x is the num of ops required to make nums[jj:ii+1] palindrome

> min ops required to make an array palindrome itself is a DP-V
    - dp[ii][jj] = min ops required to make nums[ii:jj+1] palindrome
    - dp[ii][jj] = dp[ii+1][jj-1] + 1 if nums[ii]==nums[jj] else 0


*/